schema_tables: calculate_schema_digest: prevent stalls due to large m…

bhalevy opened the pull request:
…utations vector

With a large number of table the schema mutations
vector might get big enoug to cause reactor stalls when freed.

For example, the following stall was hit on
2023.1.0~rc1-20230208.fe3cc281ec73 with 5000 tables:
```
 (inlined by) ~vector at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_vector.h:730
 (inlined by) db::schema_tables::calculate_schema_digest(seastar::sharded<service::storage_proxy>&, enum_set<super_enum<db::schema_feature, (db::schema_feature)0, (db::schema_feature)1, (db::schema_feature)2, (db::schema_feature)3, (db::schema_feature)4, (db::schema_feature)5, (db::schema_feature)6, (db::schema_feature)7> >, seastar::noncopyable_function<bool (std::basic_string_view<char, std::char_traits<char> >)>) at ./db/schema_tables.cc:799
```

This change returns a mutations generator from
the `map` lambda coroutine so we can process them
one at a time, destroy the mutations one at a time, and by that, reducing memory footprint and preventing reactor stalls.

Fixes #18173

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7898/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7898/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/860/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8890/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7898/)
#### Build Details:
- Duration: 2 hr 24 min
- Builder: spider5.cloudius-systems.com

tgrabiec commented:
Does it make sense to improve this path, since it's not taken with schema-on-raft, which is now the default?

mykaul commented:
ping?

tgrabiec commented on db/schema_tables.cc:
```diff
@@ -786,40 +786,34 @@ redact_columns_for_missing_features(mutation m, schema_features features) {
+                auto& m = *mut_opt;
+                feed_hash_for_schema_digest(hash, m, features);
+                if (diff_logger.is_enabled(logging::log_level::trace)) {
+                    diff_logger.trace("Digest {} for {}, compacted={}", hash.finalize(), m, compact_for_schema_digest(m));
```
I'm not sure it's safe to finalize "hash" here, it's used afterwards.

Also, for logging we're interested in the hash just for "m", not hash accumulated so far.

bhalevy replied:
good catch

bhalevy commented:
In v2 ([5ec3ab0](https://github.com/scylladb/scylladb/commit/5ec3ab05082b55e1bfdc208d800b329cb66d47f1))
- rebased
- fixed incremental feed hash (as per https://github.com/scylladb/scylladb/pull/18174#discussion_r1566504683)
- added: "schema_tables: calculate_schema_digest: filter the key earlier"

bhalevy commented:
boost/schema_change_test passed 100 times locally

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9297/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9297/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1796/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9820/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9297/)
#### Failed Tests (1/30523):
- [test_mv_topology_change](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9297/testReport/junit/%28root%29/test_mv_topology_change/Tests___Unit_Tests___test_mv_topology_change) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_topology_change)
#### Build Details:
- Duration: 2 hr 41 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9297/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9297/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1796/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9820/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9297/)
> 
> #### Failed Tests (1/30523):
> * [test_mv_topology_change](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9297/testReport/junit/%28root%29/test_mv_topology_change/Tests___Unit_Tests___test_mv_topology_change) [🔍](https://github.com/scylladb/scylladb/issues?q=is%3Aissue+is%3Aopen+test_mv_topology_change+sort%3Aupdated-desc)
> 
> #### Build Details:
> * Duration: 2 hr 41 min
> * Builder: spider6.cloudius-systems.com

https://github.com/scylladb/scylladb/issues/18941

respinning ci...

kostja commented:
Yielding in schema or gossiper code has been known to introduce regressions in the past @kbr-scylla jfyi.
I'm not sure this patch is safe for a backport.

bhalevy commented:
> Yielding in schema or gossiper code has been known to introduce regressions in the past @kbr-scylla jfyi. I'm not sure this patch is safe for a backport.

Unsafe how?
We're already possibly yielding in `co_await db::system_keyspace::query_mutations`
If the caller needs the digest to be atomic with what `query_mutations` returns, then they should take care of synchronization at a higher level.

bhalevy commented:
- rebased

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9687/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9687/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2067/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10090/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9687/)
#### Build Details:
- Duration: 2 hr 21 min
- Builder: spider2.cloudius-systems.com

tgrabiec commented on db/schema_tables.cc:
```diff
@@ -799,11 +799,12 @@ future<table_schema_version> calculate_schema_digest(distributed<service::storag
-            auto mut = co_await unfreeze_gently(p.mut(), s);
-            auto partition_key = value_cast<sstring>(utf8_type->deserialize(mut.key().get_component(*s, 0)));
+            auto dk = dht::decorate_key(*s, p.mut().key());
+            auto partition_key = value_cast<sstring>(utf8_type->deserialize(dk.key().get_component(*s, 0)));
```
Looks like you don't need to decorate the key.

bhalevy replied:
done in v3

bhalevy commented:
In v3 ([7dfb49e](https://github.com/scylladb/scylladb/commit/7dfb49ed08f7be38e731f8541de68cc67f0cfe78)):
- rebased
- do not decorate_key in `calculate_schema_digest`

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10224/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10224/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2468/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10493/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10224/)
#### Failed Tests (2/31109):
- [test_drop_table](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10224/testReport/junit/lwt_destructive_ddl_test/TestLwtDestructiveDDL/Tests___dtest___test_drop_table) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_drop_table)
- [test_drop_table](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10224/testReport/junit/lwt_destructive_ddl_test/TestLwtDestructiveDDL/Tests___dtest___test_drop_table_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_drop_table)
#### Build Details:
- Duration: 2 hr 12 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10224/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10224/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2468/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10493/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10224/)
> 
> #### Failed Tests (2/31109):
> * [test_drop_table](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10224/testReport/junit/lwt_destructive_ddl_test/TestLwtDestructiveDDL/Tests___dtest___test_drop_table) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_drop_table)
> * [test_drop_table](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10224/testReport/junit/lwt_destructive_ddl_test/TestLwtDestructiveDDL/Tests___dtest___test_drop_table_2) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_drop_table)

https://github.com/scylladb/scylladb/issues/19620

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10234/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10234/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10504/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2479/)
#### Failed Tests (1/5699):
- [test_repair_kill_3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10234/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_kill_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_kill_3)
#### Build Details:
- Duration: 1 hr 47 min
- Builder: spider2.cloudius-systems.com

mykaul commented:
@bhalevy - should we revive this?

bhalevy commented:
> @bhalevy - should we revive this?

Yes.  I'll rebase and retrigger CI

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10927/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10927/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11080/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3057/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10927/)
#### Build Details:
- Duration: 2 hr 15 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
@tgrabiec please re-review.
If there are no objections, let's queue this PR.
